src/server/registry/tag/tag.go

```diff
@@ -15,22 +15,165 @@
 package tag
 
 import (
+	"encoding/json"
+	"errors"
+	ierror "github.com/goharbor/harbor/src/internal/error"
+	"github.com/goharbor/harbor/src/pkg/q"
+	"github.com/goharbor/harbor/src/pkg/repository"
+	"github.com/goharbor/harbor/src/pkg/tag"
+	reg_error "github.com/goharbor/harbor/src/server/registry/error"
+	"github.com/goharbor/harbor/src/server/registry/util"
+	"github.com/gorilla/mux"
 	"net/http"
+	"sort"
+	"strconv"
 )
 
 // NewHandler returns the handler to handle listing tag request
-func NewHandler() http.Handler {
-	return &handler{}
+func NewHandler(repoMgr repository.Manager, tagMgr tag.Manager) http.Handler {
+	return &handler{
+		repoMgr: repoMgr,
+		tagMgr:  tagMgr,
+	}
 }
 
-type handler struct{}
+type handler struct {
+	repoMgr repository.Manager
+	tagMgr  tag.Manager
 
+	repositoryName string
+}
+
+// ServeHTTP ...
 func (h *handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	// TODO implement
-	// support pagination, this is required for OCI registry spec
+	switch req.Method {
+	case http.MethodGet:
+		h.get(w, req)
+	}
+}
+
+// get return the list of tags
+
+//Content-Type: application/json
+//Link: <<url>?n=<n from the request>&last=<last tag value from previous response>>; rel="next"
+//
+//{
+//    "name": "<name>",
+//    "tags": [
+//      "<tag>",
+//      ...
+//    ]
+//}
+func (h *handler) get(w http.ResponseWriter, req *http.Request) {
+	var withPagination = true
+
+	reqQ := req.URL.Query()
+	lastEntry := reqQ.Get("last")
+	maxEntries, err := strconv.Atoi(reqQ.Get("n"))
+	if err != nil || maxEntries < 0 {
+		err := ierror.New(err).WithCode(ierror.BadRequestCode).WithMessage("the N must be a positive int type")
+		reg_error.Handle(w, req, err)
+		return
+	}
+	if maxEntries == 0 {
+		withPagination = false
+	}
+
+	var repoID int64
+	var tagNames []string
+	vars := mux.Vars(req)
+	h.repositoryName = vars["name"]
+
+	repoID, err = h.getRepoID(req)
+	if err != nil {
+		reg_error.Handle(w, req, err)
+		return
+	}
+
+	// get tags ...
+	total, tags, err := h.tagMgr.List(req.Context(), &q.Query{
+		Keywords: map[string]interface{}{
+			"repository_id": repoID,
```

Comment:
Use `RepositoryID` instead. `repository_id` is column name of the database table, it shouldn't be aware by manager level

Reply:
did